File: Week Assignment 4 using C/6_sorted_insert_cll.c
// 6. Sorted insert for Circular Linked List (maintain sorted order)
#include <stdio.h>
#include <stdlib.h>

struct Node { int data; struct Node* next; };

struct Node* sortedInsertCLL(struct Node* head, int data) {
    struct Node* n = (struct Node*)malloc(sizeof(struct Node)); n->data = data; n->next = n;
    if (!head) return n;
    struct Node *cur = head, *minNode = head;
    // handle insertion where data is smallest or largest that wraps around
    do {
        if (cur->data <= data && cur->next->data >= data) {
            // insert between cur and cur->next
            n->next = cur->next; cur->next = n; return head;
        }
        cur = cur->next;
        if (cur->data < minNode->data) minNode = cur;
    } while (cur != head);
    // didn't find a middle spot: insert after max (i.e., before minNode)
    // find max node
    cur = minNode;
    while (cur->next != minNode) cur = cur->next;
    n->next = minNode; cur->next = n;
    // if new data is smaller than head, make it new head
    if (data < head->data) return n;
    return head;
}

void printCLL(struct Node* head) {
    if (!head) { printf("[]\n"); return; }
    printf("[");
    struct Node* cur = head; int first=1;
    do { if (!first) printf(", "); printf("%d", cur->data); first=0; cur=cur->next; } while (cur!=head);
    printf("]\n");
}

int main() {
    struct Node* head = NULL;
    head = sortedInsertCLL(head, 20);
    head = sortedInsertCLL(head, 10);
    head = sortedInsertCLL(head, 30);
    head = sortedInsertCLL(head, 25);
    printCLL(head); // expect [10,20,25,30]
    return 0;
}

File: Week Assignement 4 using Python/6_sorted_insert_cll.py
# 6. Sorted insert for circular linked list
class Node:
    def __init__(self,data):
        self.data=data; self.next=None

def sorted_insert(head, data):
    n = Node(data)
    if not head:
        n.next = n; return n
    cur = head
    # find spot
    while True:
        if cur.data <= data and cur.next.data >= data:
            n.next = cur.next; cur.next = n; return head
        cur = cur.next
        if cur == head: break
    # insert at end before head
    cur = head
    # find min node
    min_node = head
    while cur.next != head:
        cur = cur.next
        if cur.data < min_node.data: min_node = cur
    # insert before min_node
    cur.next = n; n.next = min_node
    if data < head.data: return n
    return head

def to_list(head):
    if not head: return []
    res=[]; cur=head
    while True:
        res.append(cur.data); cur=cur.next
        if cur==head: break
    return res

if __name__ == "__main__":
    head=None
    head=sorted_insert(head,20)
    head=sorted_insert(head,10)
    head=sorted_insert(head,30)
    head=sorted_insert(head,25)
    print(to_list(head))
